src/Data/GADT/Internal.hs

```diff
@@ -108,6 +81,9 @@ type GRead :: (k -> Type) -> Constraint
 class GRead t where
     greadsPrec :: Int -> GReadS t
 
+-- (forall a. Read (t a)) =>
```

Comment:
Really should be 
```suggestion
-- (exists a. Read (t a)) =>
```
`GRead ((:~:) a)` only reads `a :~: a`, we're not going to create non-trivial `:~:` out of nothing. Likewise, in #45 we have
```haskell
instance forall k1 k2 (a :: k1). k1 ~ k2 => GRead ((:~~:) a :: k2 -> Type)
```
which is more clearly written (at the cost of less back compat) as 
```haskell
instance forall k1 k2 (a :: k1). k1 ~ k2 => GRead ((:~~:) @k1 @k2 a)
```
We can only attempt to read when the kinds are the same and there isn't, in fact, any heterogeneity.